idle_sweep_test.go

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannel_test
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	. "github.com/uber/tchannel-go"
+
+	"github.com/uber/tchannel-go/raw"
+	"github.com/uber/tchannel-go/testutils"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func numConnections(ch *Channel) int {
+	rootPeers := ch.RootPeers().Copy()
+	count := 0
+
+	for _, peer := range rootPeers {
+		in, out := peer.NumConnections()
+		count += in + out
+	}
+
+	return count
+}
+
+func connectionStatus(channels []*Channel) string {
+	status := make([]string, 0)
+	for _, ch := range channels {
+
+		status = append(status,
+			fmt.Sprintf("%s: %d open", ch.PeerInfo().ProcessName, numConnections(ch)))
+	}
+
+	return strings.Join(status, ", ")
+}
+
+func waitForZeroConnections(t *testing.T, channels ...*Channel) bool {
```

Comment:
we could actually events here:
```
    // OnPeerStatusChanged is an optional callback that receives a notification
    // whenever the channel establishes a usable connection to a peer, or loses
    // a connection to a peer.
    OnPeerStatusChanged func(*Peer)
```

consider setting up events so we know when there's a connection change rather than waiting in a loop

Reply:
Done